src/kde/kde.t.hh

```diff
@@ -0,0 +1,356 @@
+        const double x0 = global_x_position[i];
+        // fetch local contribution
+        for (int j = 0; j<local_size; j++){
+            const double h = band_width[j][0];
+            const double x = position[j][0];
+            const double u = (x0-x)/h;
+            const double uxl = ((x-global_min_x)+(x0-global_min_x))/h;
+            const double uxu = ((global_max_x-x)+(global_max_x-x0))/h;
+            normal[j] += (epan_kernel(u)+epan_kernel(uxl)+epan_kernel(uxu))/h;
+        }
+    }
+    // fetch upper ranks nonlocal contribution
+    for (int i = global_upper_bound; i<size; i++){
+        const double x0 = global_x_position[i];
+        // fetch local contribution
+        for (int j = 0; j<local_size; j++){
+            const double h = band_width[j][0];
+            const double x = position[j][0];
+            const double u = (x0-x)/h;
+            const double uxl = ((x-global_min_x)+(x0-global_min_x))/h;
+            const double uxu = ((global_max_x-x)+(global_max_x-x0))/h;
+            normal[j] += (epan_kernel(u)+epan_kernel(uxl)+epan_kernel(uxu))/h;
+        }
+    }
+
+    // now apply the kernel to the local ranks
+    for (int i = 0; i<local_size; i++){
+        const double x0 = position[i][0];
+        // fetch local contribution
+        for (int j = 0; j<local_size; j++){
+            const double x = position[j][0];
+            const double h = band_width[j][0];
+            const double u = (x0-x)/h;
+            const double uxl = ((x-global_min_x)+(x0-global_min_x))/h;
+            const double uxu = ((global_max_x-x)+(global_max_x-x0))/h;
+            result[i] += distribution[j]*(epan_kernel(u)+epan_kernel(uxl)+epan_kernel(uxu))/h/normal[j];
+        }
+    }
+    // apply contribution to lower ranks
+    for (int i = 0; i<global_lower_bound; i++){
+        const double x0 = global_x_position[i];
+        // fetch local contribution
+        for (int j = 0; j<local_size; j++){
+            const double h = band_width[j][0];
+            const double x = position[j][0];
+            const double u = (x0-x)/h;
+            const double uxl = ((x-global_min_x)+(x0-global_min_x))/h;
+            const double uxu = ((global_max_x-x)+(global_max_x-x0))/h;
+            global_contribution[i] += distribution[j]*(epan_kernel(u)+epan_kernel(uxl)+epan_kernel(uxu))/h/normal[j];
+        }
+    }
+    // apply contribution to upper ranks
+    for (int i = global_upper_bound; i<size; i++){
+        const double x0 = global_x_position[i];
+        // fetch local contribution
+        for (int j = 0; j<local_size; j++){
+            const double h = band_width[j][0];
+            const double x = position[j][0];
+            const double u = (x0-x)/h;
+            const double uxl = ((x-global_min_x)+(x0-global_min_x))/h;
+            const double uxu = ((global_max_x-x)+(global_max_x-x0))/h;
+            global_contribution[i] += distribution[j]*(epan_kernel(u)+epan_kernel(uxl)+epan_kernel(uxu))/h/normal[j];
```

Comment:
That is a lot of repetition.  We probably should find a way to eliminate 5 of these copies.  (But I don't have a solution at the moment, although a file-scope function might work.)

Reply:
Yes, I was thinking the same. But HOW! I will ponder this. 🧐